gym_ignition/base/robot/robot_joints.py

```diff
@@ -0,0 +1,361 @@
+            velocity_reference: The joint velocity reference.
+
+        Returns:
+            True if successful, False otherwise.
+        """
+
+    @abstractmethod
+    def set_joint_force(self, joint_name: str, force: float, clip: bool = False) -> bool:
+        """
+        Set the generalized force applied to the specified joint.
+
+        Args:
+            joint_name: The name of the joint.
+            force: The joint generalize force.
+            clip: Clip joint forces reading effort limits.
+
+        Returns:
+            True if successful, False otherwise.
+        """
+
+    @abstractmethod
+    def set_joint_position(self, joint_name: str, position: float) -> bool:
+        """
+        Set the position of the specified joint.
+
+        This method attempts to achieve a perfect trajectory following. Use with care.
+        On simulated robots this method uses high-gain PID control.
+        On real robots this method bypasses all the trajectory generators that might
+        filter the reference, sending it directly to the lowest-level accessible PID.
+
+        Args:
+            joint_name: The name of the joint.
+            position: The position of the joint.
+
+        Returns:
+            True in successful, False otherwise.
+        """
+
+    @abstractmethod
+    def set_joint_velocity(self, joint_name: str, velocity: float) -> bool:
+        """
+        Set the velocity of the specified joint.
+
+        This method attempts to achieve a perfect trajectory following. Use with care.
+        On simulated robots this method uses high-gain PID control.
+        On real robots this method bypasses all the trajectory generators that might
+        filter the reference, sending it directly to the lowest-level accessible PID.
+
+        Args:
+            joint_name: The name of the joint.
+            velocity: The velocity of the joint.
+
+        Returns:
+            True in successful, False otherwise.
+        """
+
+    @abstractmethod
+    def set_joint_pid(self, joint_name: str, pid: PID) -> bool:
+        """
+        Set the low-level PID associated to the specified joint.
+
+        It can be either a position or a velocity PID.
+        Calling :meth:`set_joint_position_target` or :meth:`set_joint_position_target`
+        automatically enables position or velocity control, respectively.
+
+        Args:
+            joint_name: The name of the joint.
+            pid: The joint PID.
+
+        Returns:
+            True in successful, False otherwise.
+        """
+
+    @abstractmethod
+    def reset_joint(self,
+                    joint_name: str,
+                    position: float = None,
+                    velocity: float = None) -> bool:
+        """
+        Reset the position and velocity of a joint.
+
+        On simulated robot this method bypasses the physics.
+        On real robots this method has the same effect of set_joint_position and
+        get_joint_position.
```

Comment:
In which sense  it has the same effect of `get_joint_position`? Furthermore, if `reset_joint` is meant to be called for example with joint positions really far from the current ones, setting the low-level PID controler to a far joint position could be quite dangereous. Furthermore, is `velocity` argument ignored on real robots?

Reply:
I think we should discourage using this method on real robots. It is actually a simulator-specific method, as other ones. I didn't want to create a new interface to collect methods for simulated robots only, since using them would affect the generality we aim for Task implementations.

I agree that on real robots it would be safer to use `set_joint_position_target`. And, yes, I am not really sure how to deal with the velocity. The usage of this method is not really intuitive, do you have any recommendation about how to proceed?

This is the first iteration of the interfaces and I think we still have work to do. As soon as we'll have a working experiment that uses most of the features I think we can start thinking about all the cases, including the ones involving real robots.

Reply:
I see, then I would simply discourage to call this method in sim, or at least not to include any description on how it should be implemented on real robots.

Reply:
Ok thanks, I will update the method documentation accordingly.

Reply:
Actually, since calling this method is often related to the `gym.Env.reset()` phase, the implementation on real robot should take care of the environment reset.

This might mean also pausing the execution, ask for user manual intervention (move the robot manually, if required), and enter input when ready to continue.